Implemented the H-Measure metric for binary classification.

borjar opened the pull request:
Hi,

I added an implementation of the H-Measure, an alternative performance metric to the AUC for binary classifiers.

The AUC uses different misclassification costs (false positive - false negative) for different classifiers, which means it effectively compares different things when used on two different classifiers even on the same problem/dataset (see the following paper for proof: http://engr.case.edu/ray_soumya/mlrg/measuring_performance_hand.mlj09.pdf). The H-Measure solves this issue by explicitly specifying the relative costs of false positives and false negatives as a probability distribution.

This new method uses a beta distribution with parameters alpha and beta for the relative misclassification costs, and its implementation follows the paper mentioned above. If the parameters alpha and beta are not provided then the default distribution will be the one explained in the following paper:
http://arxiv.org/pdf/1202.2564v2.pdf

This new method was added to the metrics module in ranking.py. More information on the H-Measure can be found at the following website: http://www.hmeasure.net/

Kind regards,

 Borja

borjar committed:
Implemented the H-Measure metric for binary classification. http://www.hmeasure.net/

amueller commented:
For 1. there are quite some references, right? I haven't really heard about it before, but it seems to have been evaluated a fair bit. Let's hear from more people. If none of the devs heard of it I guess it's out ;)

borjar committed:
Added test cases for the H Measure, and checks on the input vectors.

borjar commented:
Hi, sorry for the late response,

Regarding 1, besides papers interpreting the H-Measure and comparing it to other measures, I have found the following papers that actually use it to report the results of their experiments:
- https://lirias.kuleuven.be/bitstream/123456789/270068/1/mcs_mlj2010_final_version.pdf
- http://arxiv.org/ftp/arxiv/papers/1112/1112.1668.pdf
- http://arxiv.org/ftp/arxiv/papers/1112/1112.1670.pdf
I do agree that scikit-learn should not become a platform for testing new metrics, but this one has already been around for some time and I believe there is currently no other metric in scikit-learn that overcomes the issues of the AUC in the context of binary classification in the way the H-Measure does. I think it would be a reasonable addition to the project.

Regarding 2, I have added the h_measure_score() function to the metrics/ranking.py module where the other ROC/AUC related functions are. Although based on it, the metric is not just a weighting over the ROC, I'm not sure how it could be further integrated within the current ROC framework, but I'd be happy to modify it if you had some specific suggestions.

Regarding 3, I have added the test cases in metrics/tests/test_ranking.py. Is it enough to add a test function in there, or do I need to reference it somewhere?

Kind regards,

Borja

amueller commented:
For 3: The tests are run automatically if they star with ``test_``

amueller commented:
Did you look at travis? The doctest is failing.

borjar committed:
Fixed the output of the example in the docstring of h_measure_score().

borjar commented:
Hi,

Sorry I wasn't sure what the Travis CI build was. I've fixed the error and updated the code.

Borja

jnothman commented:
> For 1. there are quite some references, right? I haven't really heard about it before, but it seems to have been evaluated a fair bit.

Is that sufficient criteria of inclusion for a metric? I suggest it would be good to see more work using it first. The cited papers using it are not really influential works and may have chosen the metric as "cool and new"... I would be more convinced if we found a continuity of works using the measure from 2010 until now.

amueller commented:
On 01/20/2015 06:58 PM, jnothman wrote:
>
>     For 1. there are quite some references, right? I haven't really
>     heard about it before, but it seems to have been evaluated a fair bit.
>
> Is that sufficient criteria of inclusion for a metric? I suggest it 
> would be good to see more work using it first. The cited papers using 
> it are not really influential works and may have chosen the metric as 
> "cool and new"... I would be more convinced if we found a continuity 
> of works using the measure from 2010 until now.
>
Yeah, that would be definitely better.
I'm note entirely sure how high we should set the bar.
To answer your first question: probably not.

thomasjpfan commented:
Looking at the references in 2022 they are on the order of ~ 50 citations, thus the feature does not meet [scikit-learn's inclusion criterion](https://scikit-learn.org/dev/faq.html#what-are-the-inclusion-criteria-for-new-algorithms). With that in mind, I am closing this PR.
